---
title: "MINOTAUR Workflow Vignette"
author: "Daren C. Card"
date: "December 11, 2015"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{MINOTAUR Workflow Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Introduction
MINOTAUR is a R package for identifying and visualizing multivariate outliers in general, but is tailored towards researchers seeking to identify genetic loci that are under natural selection or that are putative underlying phenotypic traits. Methodologies targeting these answers typically use whole-genome sequencing or reduced marker datasets from population-level sampling and infer one or more meaningful statistics, including population genetic statistics, genotype by environment statistics, and genome-wide association statistics. In effect, generating these statistics from loci spread throughout a genome allows a geneticist to scan for loci or regions of the genome where a statistic behaves aberantly in comparison with the genome-wide background. 

Studies are increasingly generating several statistics across one or more broad inferential classes, and a common practice thus far has been to look across statistics to identify candidate genes for further investigation, essentially enriching for loci with outlier values for several statistics. The largely arbitrary nature of this decision making process and the fact that certain statistics enjoy more power under particular evolutionary scenarios or sampling designs collectively indicate that these methods can be biased and can suffer from high error rates. MINOTAUR was designed to help identify candidate genomic loci or regions by leveraging values from several statistics to identify outliers in multi-dimensional space. The core goal of this package is to aid in outlier visualization using the R Shiny interface, and secondarily it also introduce new multivariate outlier detection procedures to supplement those already in existence. 

## Getting Started
MINOTAUR can be installed from [CRAN](http://cran.rstudio.com/package=XXX), or from [GitHub](http://github.com/NESCent/MINOTAUR), if the latest development version is desired:

```{r eval=FALSE}
install.packages('MINOTAUR')
# to install the development version, run:
# require(devtools)
# install_github('NESCent/MINOTAUR')
```

## Workflow Overview
The MINOTAUR package is broken into two fairly autonomous parts:

1. Four functions that perform multivariate outlier detection given data for several statistics. The multivariate outlier functions are meant to supplement existing R functions that infer outliers in multidimensional space, which all work using data stored as a dataframe object.

2. An R Shiny application that can be used to visualize the distributions of individual statictistics separately and together, and to help produce publication-quality figures. The Shiny app can be used on unfiltered data to gain baseline understanding of a dataset before multivariate outliers are inferred, and can also be used on a reduce dataset of genomic loci or regions that are inferred to be multivariate outliers.

An expanded example with code is included below, but a basic workflow leveraging MINOTAUR is as follows:

1. Data is loaded into R from an external data format (e.g., tab-delimited or comma-delimited files).

2. The MINOTAUR Shiny app can be used to perform a preliminary investigation of the data, such that informed decisions can be made as to which statistics to include in multivariate outlier analyses. Users can also use the app to exclude and filter data in preparation for the multivariate analyses, which cannot accept missing data.

3. Outside of the Shiny app, a user can infer multivariate outliers using the statistics of interest and functions provided as part of MINOTAUR and several other packages in the R environment.

4. The filtered dataset with only multivariate outliers can be visualized using the MINOTAUR Shiny app, which enables further investigation and the production of publication-quality figures.

## Quick Start
### Inferring Multivariate Outliers
Get going and understand the basics of inferring multivariate outliers by running the following code using the sample data distributed with MINOTAUR.
```{r eval=FALSE}
data <- read.table("XXX", header=TRUE)              # read the data into R, dataframe conatins observations in rows and statistics in columns
columns <- c(4,5,6,7)                               # columns of statistics to be used for the multivariate outlier analyses
Md <- Mahalanobis(data, columns)
Hd <- harmonicDist(data, columns)                   # all MINOTAUR functions take two standard arguments
Nd <- neighborDist(data, columns)
bandwidth <- c(seq(0.01, 0.1, by=0.01),             # the kernel density function takes a third argument: the kernel bandwidth
  seq(0.2, 1, by=0.1))
Kd.ML <- kernelLogLike(data, columns, bandwidth)    # a function has been provided that estimates the maximum likelihood kernel bandwidth that should be used based on the data
bestBandwidth <- bandwidth[which(Kd.ML==max(Kd.ML))[1]]
Kd <- kernelDist(data, columns, bestBandwidth)      # this bandwidth can then be used to run the kernel density function
data$Md <- Md
data$Hd <- Hd                                       # multivariate results can be appended to the input data matrix for visualization
data$Nd <- Nd
data$Kd <- Kd
```

### Data Visualization and Manipulation in R Shiny
The graphical nature of the Shiny application should make this portion of the package intuitive, and we've provided some annotated snapshots of the app to aid the beginning user.



## Specific Example



## Acknowledgements
This work was conceived during a hackathon on Population Genetics in R, sponsored and hosted by the National Evolutionary Synthesis Center (NESCent) in March of 2015. The MINOTAUR team is grateful to fellow hackathon participants who gave valuable feedback during initial development of this project at NESCent and afterwards. We are also indebted to those at NESCent who made the event possible and especially to Hilmar Lapp, who organized and led the collective effort.

## Template Vignette (for reference only)

Vignettes are long form documentation commonly included in packages. Because they are part of the distribution of the package, they need to be as compact as possible. The `html_vignette` output type provides a custom style sheet (and tweaks some options) to ensure that the resulting html is as small as possible. The `html_vignette` format:

- Never uses retina figures
- Has a smaller default figure size
- Uses a custom CSS stylesheet instead of the default Twitter Bootstrap style

## Vignette Info

Note the various macros within the `vignette` section of the metadata block above. These are required in order to instruct R how to build the vignette. Note that you should change the `title` field and the `\VignetteIndexEntry` to match the title of your vignette.

## Styles

The `html_vignette` template includes a basic CSS theme. To override this theme you can specify your own CSS in the document metadata as follows:

    output: 
      rmarkdown::html_vignette:
        css: mystyles.css

## Figures

The figure sizes have been customised so that you can easily put two images side-by-side. 

```{r, fig.show='hold'}
plot(1:10)
plot(10:1)
```

You can enable figure captions by `fig_caption: yes` in YAML:

    output:
      rmarkdown::html_vignette:
        fig_caption: yes

Then you can use the chunk option `fig.cap = "Your figure caption."` in **knitr**.

## More Examples

You can write math expressions, e.g. $Y = X\beta + \epsilon$, footnotes^[A footnote here.], and tables, e.g. using `knitr::kable()`.

```{r, echo=FALSE, results='asis'}
knitr::kable(head(mtcars, 10))
```

Also a quote using `>`:

> "He who gives up [code] safety for [code] speed deserves neither."
([via](https://twitter.com/hadleywickham/status/504368538874703872))
